vespaclient-container-plugin/src/main/java/com/yahoo/document/restapi/DocumentOperationExecutor.java

```diff
@@ -0,0 +1,598 @@
+                    throttled.notify();
+                }
+                break;
+            default:
+                log.log(Level.WARNING, "Unknown result type '" + result.type() + "'");
+            case FATAL_ERROR: // intentional fallthrough
+                context.error(ERROR, result.getError().getMessage());
+        }
+    }
+
+    private void handleResponse(Response response){
+        OperationContext context = outstanding.remove(response.getRequestId());
+        if (context != null)
+            if (response.isSuccess())
+                context.success(response instanceof DocumentResponse ? Optional.ofNullable(((DocumentResponse) response).getDocument())
+                                                                     : Optional.empty());
+            else
+                context.error(toErrorType(response.outcome()), response.getTextMessage());
+        else
+            log.log(Level.FINE, () -> "Received response for request which has timed out, with id " + response.getRequestId());
+    }
+
+    private static ErrorType toErrorType(Response.Outcome outcome) {
+        switch (outcome) {
+            case NOT_FOUND:
+                return NOT_FOUND;
+            case CONDITION_FAILED:
+                return PRECONDITION_FAILED;
+            default:
+                log.log(Level.WARNING, "Unexpected response outcome: " + outcome);
+            case ERROR: // intentional fallthrough
+                return ERROR;
+        }
+    }
+
+    private void retry() {
+        maintain(throttled, delayed -> send(delayed.operation(), delayed.context()));
+    }
+
+    private void timeOut() {
+        maintain(timeouts, delayed -> delayed.context().error(TIMEOUT, "Timed out after " + defaultTimeout));
+    }
+
+    private void maintain(ConcurrentLinkedQueue<Delayed> queue, Consumer<Delayed> action) {
+        while ( ! Thread.currentThread().isInterrupted()) {
+            try {
+                long waitUntilMillis = Instant.MAX.toEpochMilli();
+                Iterator<Delayed> operations = queue.iterator();
+                while (operations.hasNext()) {
+                    Delayed operation = operations.next();
+                    // Already handled: remove and continue.
+                    if (operation.context().handled()) {
+                        operations.remove();
+                        continue;
+                    }
+                    // Ready for retry — remove from queue and send.
+                    if (operation.readyAt().isBefore(clock.instant())) {
+                        action.accept(operation);
+                        operations.remove();
+                        continue;
+                    }
+                    // Not yet ready for retry — track earliest time to wake up again.
+                    waitUntilMillis = Math.min(waitUntilMillis, operation.readyAt.toEpochMilli());
+                }
+                synchronized (queue) {
```

Comment:
Do not use a concurrent collection as an external monitor (unless `ConcurrentLinkedQueue` says it's okay).

Reply:
Good point. Already changed to allow tests to notify maintainers, though :)